File: skills/wiggle_joint/wiggle_joint.cc
#include "wiggle_joint.h"

#include <memory>
#include <string>
#include <utility>

#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "skills/wiggle_joint/wiggle_joint.pb.h"
// [START wiggle_joint_includes]
#include "intrinsic/icon/actions/point_to_point_move_info.h"
#include "intrinsic/icon/cc_client/client.h"
#include "intrinsic/icon/cc_client/client_utils.h"
#include "intrinsic/icon/cc_client/session.h"
#include "intrinsic/icon/equipment/channel_factory.h"
#include "intrinsic/icon/equipment/equipment_utils.h"
// [END wiggle_joint_includes]
#include "intrinsic/util/status/status_macros.h"
#include "intrinsic/skills/cc/skill_utils.h"
#include "intrinsic/skills/proto/skill_service.pb.h"

namespace wiggle_joint {

using ::com::example::WiggleJointParams;

// [START wiggle_joint_using_directives_p1]
using ::intrinsic_proto::icon::PartJointState;
using ::intrinsic_proto::icon::PartStatus;
// [END wiggle_joint_using_directives_p1]
using ::intrinsic::skills::ExecuteRequest;
using ::intrinsic_proto::skills::ExecuteResult;
// [START wiggle_joint_using_directives_p2]
using ::intrinsic::icon::Action;
using ::intrinsic::icon::ActionDescriptor;
using ::intrinsic::icon::ActionInstanceId;
using ::intrinsic::icon::ConnectToIconEquipment;
using ::intrinsic::icon::CreatePointToPointMoveFixedParams;
using ::intrinsic::icon::DefaultChannelFactory;
using ::intrinsic::icon::IconEquipment;
using ::intrinsic::icon::IsDone;
using ::intrinsic::icon::IsGreaterThanOrEqual;
using ::intrinsic::icon::IsTrue;
using ::intrinsic::icon::PointToPointMoveInfo;
using ::intrinsic::icon::ReactionDescriptor;
using ::intrinsic::skills::EquipmentPack;
// [END wiggle_joint_using_directives_p2]
using ::intrinsic::skills::SkillInterface;
using ::intrinsic::skills::ExecuteContext;

// [START wiggle_joint_using_directives_p3]
using IconClient = ::intrinsic::icon::Client;
using IconSession = ::intrinsic::icon::Session;
// [END wiggle_joint_using_directives_p3]
// -----------------------------------------------------------------------------
// Skill signature.
// -----------------------------------------------------------------------------

std::unique_ptr<SkillInterface> WiggleJoint::CreateSkill() {
  return std::make_unique<WiggleJoint>();
}

// -----------------------------------------------------------------------------
// Skill execution.
// -----------------------------------------------------------------------------

absl::StatusOr<std::unique_ptr<google::protobuf::Message>> WiggleJoint::Execute(
    const ExecuteRequest& request, ExecuteContext& context) {

  // Get parameters.
  INTR_ASSIGN_OR_RETURN(
    auto params, request.params<WiggleJointParams>());

  // [START access_equipment]
  const EquipmentPack equipment_pack = context.equipment();
  INTR_ASSIGN_OR_RETURN(const auto equipment, equipment_pack.GetHandle(kRobotSlot));
  // [END access_equipment]
  // [START connect_to_robot]
  INTR_ASSIGN_OR_RETURN(IconEquipment icon_equipment,
    ConnectToIconEquipment(equipment_pack, kRobotSlot, DefaultChannelFactory()));
  IconClient icon_client(icon_equipment.channel);
  // [END connect_to_robot]
  // [START get_robot_info]
  INTR_ASSIGN_OR_RETURN(auto robot_config, icon_client.GetConfig());

   const std::string part_name = icon_equipment.position_part_name.value();
  INTR_ASSIGN_OR_RETURN(intrinsic_proto::icon::GenericPartConfig part_config,
                   robot_config.GetGenericPartConfig(part_name));

  INTR_ASSIGN_OR_RETURN(const PartStatus part_status,
                   icon_client.GetSinglePartStatus(part_name));
  // [END get_robot_info]
  // [START parameter_validation]
  if (
    params.joint_number() < 0 ||
    params.joint_number() >= part_status.joint_states_size()) {

    return absl::FailedPreconditionError(absl::StrCat(
      "Joint ", params.joint_number(), " does not exist on: ",
      equipment.name()));
  }
  // [END parameter_validation]
  // [START describe_movement_goals]
  const double five_degrees = 0.0872665;  // Radians
  const double joint_target_pose =
      part_status.joint_states(
        params.joint_number()).position_sensed() + five_degrees;

  std::vector<double> original_positions;
  std::vector<double> first_goal_positions;
  std::vector<double> zero_velocity;
  for (size_t i = 0; i < part_status.joint_states_size(); ++i) {
      const double cur_pos = part_status.joint_states(i).position_sensed();
      if (params.joint_number() == i) {
          first_goal_positions.push_back(joint_target_pose);
      } else {
          first_goal_positions.push_back(cur_pos);
      }
      original_positions.push_back(cur_pos);
      zero_velocity.push_back(0);
  }
  // [END describe_movement_goals]

  // [START start_icon_session]
  // Start an icon session
  INTR_ASSIGN_OR_RETURN(
    std::unique_ptr<IconSession> icon_session,
    IconSession::Start(icon_equipment.channel, {std::string(part_name)}));
  // [END start_icon_session]

  // [START icon_constants]
  const ActionInstanceId kFirstMoveId(1);
  const ActionInstanceId kSecondMoveId(2);
  const double kTimeoutSeconds = 2;
  // [END icon_constants]

  // [START first_icon_action]
  ActionDescriptor first_move =
    // [START first_icon_action_id_and_goal]
    ActionDescriptor(PointToPointMoveInfo::kActionTypeName,
                            kFirstMoveId, part_name)
        .WithFixedParams(CreatePointToPointMoveFixedParams(
            first_goal_positions, zero_velocity))
    // [END first_icon_action_id_and_goal]
    // [START first_icon_action_when_done]
        .WithReaction(
            ReactionDescriptor(IsDone())
            .WithRealtimeActionOnCondition(kSecondMoveId))
    // [END first_icon_action_when_done]
    // [START first_icon_action_when_timeout]
        .WithReaction(
            ReactionDescriptor(
                IsGreaterThanOrEqual(
                          PointToPointMoveInfo::kSetpointDoneForSeconds,
                          kTimeoutSeconds))
            // This nonrealtime callback prints a message if the
            // joint motion fails to settle within a timeout.
            .WithWatcherOnCondition([&icon_session]() {
              LOG(ERROR) << "Failed to reach Goal";
              icon_session->QuitWatcherLoop();
            }));
    // [END first_icon_action_when_timeout]
  // [END first_icon_action]
  // [START second_icon_action]
  ActionDescriptor second_move =
    ActionDescriptor(PointToPointMoveInfo::kActionTypeName,
                            kSecondMoveId, part_name)
        .WithFixedParams(CreatePointToPointMoveFixedParams(
            original_positions, zero_velocity))
        .WithReaction(
            ReactionDescriptor(
                IsTrue(PointToPointMoveInfo::kIsSettled))
            .WithWatcherOnCondition([&icon_session]() {
                LOG(INFO) << "Finished moving joint.";
                icon_session->QuitWatcherLoop();
              }))
        .WithReaction(
            ReactionDescriptor(
                IsGreaterThanOrEqual(
                    PointToPointMoveInfo::kSetpointDoneForSeconds,
                    kTimeoutSeconds))
            // This nonrealtime callback prints a message if the
            // joint motion fails to settle within a timeout.
            .WithWatcherOnCondition([&icon_session]() {
              LOG(ERROR) << "Failed to reach Goal";
              icon_session->QuitWatcherLoop();
            }));
  // [END second_icon_action]

  // [START add_icon_actions]
  INTR_ASSIGN_OR_RETURN(
    auto actions,
    icon_session->AddActions({first_move, second_move}));
  // [END add_icon_actions]

  // [START move_the_robot]
  LOG(INFO) << "Starting Wiggle.";
  INTR_RETURN_IF_ERROR(
    icon_session->StartActions({actions.front()}));
  INTR_RETURN_IF_ERROR(icon_session->RunWatcherLoop());
  LOG(INFO) << "Finished Wiggle.";
  // [END move_the_robot]

  return nullptr;
}
}  // namespace wiggle_joint

File: skills/wiggle_joint/wiggle_joint.py
from absl import logging

from intrinsic.icon.actions import point_to_point_move_pb2
from intrinsic.icon.equipment import equipment_utils
from intrinsic.icon.proto import joint_space_pb2
from intrinsic.icon.python import icon_api
from intrinsic.skills.proto import equipment_pb2
from intrinsic.skills.proto import skill_service_pb2
from intrinsic.skills.python import proto_utils
from intrinsic.skills.python import skill_interface
from intrinsic.util.decorators import overrides

from skills.wiggle_joint import wiggle_joint_pb2

ROBOT_EQUIPMENT_SLOT: str = "robot"


def get_single_part_status(status_response, part_name):
    if part_name in status_response.part_status:
        return status_response.part_status[part_name]


class WiggleJoint(skill_interface.Skill):
    """Skill that wiggles a single joint on a robot."""

    @overrides(skill_interface.Skill)
    def execute(
        self,
        request: skill_service_pb2.ExecuteRequest,
        context: skill_interface.ExecuteContext,
    ) -> None:

        # Get equipment.
        icon_equipment: equipment_pb2.EquipmentHandle = (
            context.resource_handles[ROBOT_EQUIPMENT_SLOT]
        )

        icon_client = equipment_utils.init_icon_client(icon_equipment)

        part_name = equipment_utils.get_position_part_name(icon_equipment)
        part_status = get_single_part_status(icon_client.get_status(), part_name)

        if part_status is None:
            raise ValueError(f"Could not get status for {part_name}")

        indexTooSmall = request.params.joint_number < 0
        indexTooBig = request.params.joint_number >= len(part_status.joint_states)
        if indexTooSmall or indexTooBig:
            raise ValueError(
                f"Joint {request.params.joint_number} does not exist"
                f" on {icon_equipment.name}"
            )

        five_degrees = 0.0872665  # Radians
        joint_target_pose = (
            part_status.joint_states[request.params.joint_number].position_sensed + five_degrees
        )

        original_positions = joint_space_pb2.JointVec()
        first_goal_positions = joint_space_pb2.JointVec()
        zero_velocity = joint_space_pb2.JointVec()
        for i, cur_state in enumerate(part_status.joint_states):
            cur_pos = cur_state.position_sensed
            if i == request.params.joint_number:
                first_goal_positions.joints.append(joint_target_pose)
            else:
                first_goal_positions.joints.append(cur_pos)
            original_positions.joints.append(cur_pos)
            zero_velocity.joints.append(0)

        FIRST_ACTION_ID = 1
        SECOND_ACTION_ID = 2

        TIMEOUT_SECONDS = 2.0

        sig_actions_done = icon_api.SignalFlag()
        success = True

        def on_timeout(timestamp, prev_action_id, current_action_id):
            nonlocal success
            success = False
            logging.error(f"Action {current_action_id} timed out")

        with icon_client.start_session([part_name]) as icon_session:
            first_move = icon_api.Action(
                action_id=FIRST_ACTION_ID,
                action_type="xfa.point_to_point_move",
                part_name_or_slot_part_map=part_name,
                params=point_to_point_move_pb2.PointToPointMoveFixedParams(
                    goal_position=first_goal_positions,
                    goal_velocity=zero_velocity,
                ),
                reactions=[
                    icon_api.Reaction(
                        condition=icon_api.Condition.is_true("xfa.is_settled"),
                        responses=[
                            icon_api.StartActionInRealTime(
                                start_action_id=SECOND_ACTION_ID,
                            )
                        ],
                    ),
                    icon_api.Reaction(
                        condition=icon_api.Condition.is_greater_than_or_equal(
                            state_variable_name="xfa.setpoint_done_for_seconds",
                            value=TIMEOUT_SECONDS,
                        ),
                        responses=[
                            icon_api.TriggerCallback(on_timeout),
                            icon_api.Signal(sig_actions_done),
                        ],
                    ),
                ],
            )
            second_move = icon_api.Action(
                action_id=SECOND_ACTION_ID,
                action_type="xfa.point_to_point_move",
                part_name_or_slot_part_map=part_name,
                params=point_to_point_move_pb2.PointToPointMoveFixedParams(
                    goal_position=original_positions,
                    goal_velocity=zero_velocity,
                ),
                reactions=[
                    icon_api.Reaction(
                        condition=icon_api.Condition.is_true("xfa.is_settled"),
                        responses=[icon_api.Signal(sig_actions_done)],
                    ),
                    icon_api.Reaction(
                        condition=icon_api.Condition.is_greater_than_or_equal(
                            state_variable_name="xfa.setpoint_done_for_seconds",
                            value=TIMEOUT_SECONDS,
                        ),
                        responses=[
                            icon_api.TriggerCallback(on_timeout),
                            icon_api.Signal(sig_actions_done),
                        ],
                    ),
                ],
            )
            icon_session.add_actions([first_move, second_move])
            icon_session.start_action(FIRST_ACTION_ID)

            logging.info("WiggleJoint waiting for robot to move")
            sig_actions_done.wait()
            logging.info("WiggleJoint movement is complete")

        if not success:
            raise RuntimeError("Failed to wiggle joint")
